### PR TITLE
Use DirectByteBuffer pool and submit requests while iterating

### PR DESCRIPTION
**Pull request by jordeu**
## Description 
This version of the s3 fs has five main changes:

1. Use a `DirectByteBufferPool` instead of chunk factory
2. Uses `SequenceInputStream` to join all the streams
3. Submits the request to the executor while iterating not all of them at the begging
4. On out of memory error falls back to a `DownloadOnDemandInputStream` that downloads on demand (not in parallel).
5. Reduce concurrency when the SDK requests fail on parallel.

## Performance

Original branch `s3fs-refactor`:
```
$ time NXF_S3_DOWNLOAD_PARALLEL=true ./launch.sh fs cp s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1 sample.bax
Picked up JAVA_TOOL_OPTIONS:  -Xmx2576m

real    11m36.814s
user    67m43.463s
sys     0m15.589s
gitpod /workspace/nextflow $ md5sum sample.bax 
af25b225101baf29598269eb358991a0  sample.bax
```

This branch `s3fs-refactor-alt`:
```
$ time NXF_S3_DOWNLOAD_PARALLEL=true ./launch.sh fs cp s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1 sample.bax
Picked up JAVA_TOOL_OPTIONS:  -Xmx2576m

real    1m16.725s
user    6m45.221s
sys     0m11.658s
gitpod /workspace/nextflow $ md5sum sample.bax 
af25b225101baf29598269eb358991a0  sample.bax
```

### CODE REVIEW

**Comment by jordeu** [comment]
Example forcing an OOM error:

```
$ time NXF_S3_DOWNLOAD_CHUNK_SIZE="100 MB" NXF_S3_DOWNLOAD_NUM_WORKERS=20 NXF_S3_DOWNLOAD_PARALLEL=true ./launch.sh fs cp s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1 sample.bax
Picked up JAVA_TOOL_OPTIONS:  -Xmx2576m
WARN: Out of memory downloading chunk range=3774873600..3879731199; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
WARN: Out of memory downloading chunk range=3355443200..3460300799; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
WARN: Out of memory downloading chunk range=3879731200..3984588799; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
WARN: Out of memory downloading chunk range=3984588800..4089446399; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
WARN: Out of memory downloading chunk range=4194304000..4299161599; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
WARN: Out of memory downloading chunk range=4613734400..4699494993; path=s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1. Fallback to direct copy.
Unexpected error [NullPointerException]



real    1m55.116s
user    7m23.327s
sys     0m12.007s
gitpod /workspace/nextflow $ md5sum sample.bax 
af25b225101baf29598269eb358991a0  sample.bax
```

**Comment by jordeu** [comment]
But there is still room for improvement, because using AWS cli using also 10 workers:

```
$ time /workspace/bin/aws s3 cp s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1 sample.bax
download: s3://sra-pub-src-2/ERR1305737/m151221_184918_00127_c100958412550000001823217206251660_s1_p0.1.bax.h5.1 to ./sample.bax

real    0m49.410s
user    0m31.890s
sys     0m23.624s
gitpod /workspace/nextflow $ md5sum sample.bax 
af25b225101baf29598269eb358991a0  sample.bax
```

File: plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+                        try {
+                            Thread.sleep(1000);
+                        } catch (InterruptedException ex) {
+                            cleanUpAfterException();
+                            Thread.currentThread().interrupt();
+                        }
+                        log.warn("Downloading at main thread chunk {}; path={}.", getLogRef(task.request), getLogPath(task.request));
+                        return new DownloadOnDemandInputStream(task.request, s3Client);
+                    }
+
```
Comment (pditommaso) [inline plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java:184]: Is this really needed? The `pool` should not thrown ant exception

**Comment by pditommaso** [comment]
Merge manually by 3d0f0e036. 

Main changes: 
- Start submitting up to 2 * numWorkers tasks, then just submit a new one each time a chunk is read 
- Abstracts the future iterator into its `FutureIterator` class using generics

File: plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java (hunk 1 above)
Comment (jordeu) [reply plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java:184]: It is the download who is throwing the exception. This one comes from the `s3client.getObject` call. I'm handling it here to be able to reduce the pool max threads.

Comment (pditommaso) [reply plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java:184]: I see, then we need to add it again because I've removed it

**Pull request closed without merging**